components/region_cache_memory_engine/src/write_batch.rs

```diff
@@ -58,8 +58,14 @@ impl RangeCacheWriteBatch {
     }
 
     fn write_impl(&mut self, seq: u64) -> Result<()> {
+        self.engine.handle_pending_load();
         let (engine, filtered_keys) = {
-            let core = self.engine.core().read().unwrap();
+            let mut core = self.engine.core().write().unwrap();
+
+            self.buffer
+                .iter()
+                .for_each(|e| e.maybe_cached(seq, &mut core));
```

Comment:
for every entry to go through maybe_cached, it may be too expensive. Ideally it's done once per region.